From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Red-A-Hawk <you@example.com>
Date: Tue, 8 Jan 2026 07:30:00 +0000
Subject: [PATCH] fs/exec: add SukiSU builtin execveat manual hook

---
 fs/exec.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/fs/exec.c b/fs/exec.c
index 000000000000..111111111111 100644
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1715,6 +1715,15 @@ static int __do_execve_file(int fd, struct filename *filename,
 	struct linux_binprm *bprm;
 	struct files_struct *displaced;
 	int retval;
+
+#ifdef CONFIG_KSU
+	extern bool ksu_execveat_hook;
+	extern bool ksu_boot_completed;
+	extern int ksu_handle_execveat(
+		int *fd, struct filename **filename_ptr,
+		void *argv, void *envp, int *flags);
+#endif
+
 	retval = unshare_files(&displaced);
 	if (retval)
 		return retval;
@@ -1726,6 +1735,15 @@ static int __do_execve_file(int fd, struct filename *filename,
 		goto out_ret;
 	}
 
+#ifdef CONFIG_KSU
+	if (unlikely(ksu_execveat_hook && ksu_boot_completed)) {
+		retval = ksu_handle_execveat(
+			&fd, &filename, argv, envp, &flags);
+		if (retval)
+			goto out_ret;
+	}
+#endif
+
 	retval = exec_binprm(bprm);
 
 out_ret:
