--- a/scripts/Makefile.build
+++ b/scripts/Makefile.build
@@ -319,13 +319,13 @@
 #
 # Rule to compile a set of .o files into one .o file
 #
-ifdef builtin-target
-
 ifdef CONFIG_LTO_CLANG
   ifdef CONFIG_MODVERSIONS
     # combine symversions for later processing
     update_lto_symversions =				\
 	rm -f $@.symversions; 				\
 	for i in $(foreach n,					\
 			$(filter-out FORCE,$^),			\
 			$(if $(wildcard $(n).symversions),$(n))); do	\
 		cat $$i.symversions >> $@.symversions;		\
 	done;
   endif
   # rebuild the symbol table with llvm-ar to include IR files
   update_lto_symtable = ;				\
 	mv -f $@ $@.tmp;				\
 	$(LLVM_AR) rcsT$(KBUILD_ARFLAGS) $@ 		\
 		$$($(AR) t $@.tmp); 			\
 	rm -f $@.tmp
 endif
 
+ifdef builtin-target
+
 # built-in.a archives are made with no symbol table or index which
 # makes them small and fast, but unable to be used by the linker.
 # scripts/link-vmlinux.sh builds an aggregate built-in.a with a symbol
